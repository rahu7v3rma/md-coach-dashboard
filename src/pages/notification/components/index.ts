import NotificationCard from './notificationCard';
import NotificationItem from './notificationItem';

export { NotificationCard, NotificationItem };
