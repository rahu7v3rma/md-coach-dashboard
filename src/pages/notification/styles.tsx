import styled from 'styled-components';

import { Colors } from 'src/utils/colors';

export const Wrapper = styled.div`
    width: 100%;
    height: 100vh;
    display: flex;
    background: ${Colors.extra.white1};
`;

export const MainContent = styled.div`
    position: relative;
    flex: 0 0 80%;
`;

/* Main Notification Styling */

export const Title = styled.p`
    font-family: 'Poppins';
    font-weight: 700;
    font-size: 32px;
    line-height: 36px;
    color: ${Colors.extra.blackText};
`;

export const Container = styled.div`
    display: flex;
    flex-wrap: wrap;
    width: 100%;
    height: 100%;
`;

export const NotificationList = styled.div`
    display: flex;
    flex-direction: column;
    flex: 0 0 20%;
    max-width: 250px;
    height: 100%;
    overflow-y: auto;
    padding-right: 1em;
`;

export const NotificationContent = styled.div`
    flex: 0 0 80%;
    width: 100%;
    height: 100%;
`;

export const PlaceholderContainer = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    height: 100%;
`;

export const PlaceholderFox = styled.img`
    transform: scale(0.8);
`;

export const PlaceholderText = styled.span`
    color: ${Colors.extra.slateBlue};
    font-weight: 600;
    font-size: 20px;
    text-align: center;
    margin-top: 1em;
`;

export const NotificationWrapper = styled.div`
    padding-left: 1.5em;
`;

/** End of Main Notification Main Styling */

/** Start of Notification Item */

export const NotificationContainer = styled.div`
    display: flex;
    padding: 0.5em 0.25em;
    cursor: pointer;
    width: 100%;
    margin-bottom: 0.25em;
    box-sizing: border-box;

    &:hover {
        opacity: 0.8;
    }
`;

export const DisplayName = styled.div`
    display: flex;
    flex-direction: column;
    margin-left: 1em;
`;

export const NameTitle = styled.span`
    color: ${Colors.extra.blackText};
    font-weight: 600;
    font-size: 13px;
`;

export const Subtitle = styled.span`
    color: ${Colors.theme.gray};
    font-size: 11px;
`;

export const TimeText = styled.span`
    color: ${Colors.theme.gray};
    font-size: 11px;
    margin-left: auto;
`;
