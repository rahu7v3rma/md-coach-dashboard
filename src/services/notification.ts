import { appRouter } from '../navigation';
import { NOTIFICATION_TYPE } from '../utils/constants';

export const requestPermissions = () => {
    return Notification.requestPermission().then((permission) => {
        if (permission === 'granted') {
            return true;
        } else {
            return false;
        }
    });
};

export const hasPermissions = () => {
    return Notification.permission === 'granted';
};

export const notificationAction = (
    notificationType: string,
    notificationPayload: string | null
) => {
    if (notificationType === NOTIFICATION_TYPE.STREAM_CHAT_MESSAGE) {
        if (notificationPayload) {
            // for chat notifications the payload is a channel id
            appRouter.navigate('/chat', {
                state: { channelId: notificationPayload }
            });
        }
    } else if (
        notificationType === NOTIFICATION_TYPE.LESSON_COMPLETED &&
        notificationPayload
    ) {
        appRouter.navigate(`/client/${notificationPayload}`, {});
    }
};
