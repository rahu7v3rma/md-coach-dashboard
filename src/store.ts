import { configureStore } from '@reduxjs/toolkit';

import clientReducer from './reducers/client';
import groupReducer from './reducers/group';
import notificationReducer from './reducers/notification';
import userReducer from './reducers/user';

// define store options as a variable so we can export it for tests to use
const storeOptions: any = {
    reducer: {
        user: userReducer,
        client: clientReducer,
        notification: notificationReducer,
        group: groupReducer
    }
};

const store = configureStore(storeOptions);

export default store;

// infer the `RootState` type from the store
export type RootState = ReturnType<typeof store.getState>;

// infer the `AppDispatch` type from the store so thunk actions can return promises
export type AppDispatch = typeof store.dispatch;

export const test = {
    storeOptions
};
