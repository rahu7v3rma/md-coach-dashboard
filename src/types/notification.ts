export type NotificationListResponse = {
    list: NotificationType[];
    hasPrevious: boolean;
    hasNext: boolean;
    totalCount: number;
    nextPageNumber: number;
    previousPageNumber: number;
};

export type NotificationType = {
    id: number;
    type: string;
    title: string;
    description: string;
    date_time: string;
    read_flag: boolean;
    payload: string;
};
