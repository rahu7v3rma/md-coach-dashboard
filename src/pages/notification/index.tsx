import { FunctionComponent } from 'react';

import FoxImage from '../../assets/md-mascot.svg';
import { Card, Header, Sidebar } from '../../shared';

import { NotificationItem } from './components';
import {
    Container,
    MainContent,
    NotificationContent,
    NotificationList,
    NotificationWrapper,
    PlaceholderContainer,
    PlaceholderFox,
    PlaceholderText,
    Title,
    Wrapper
} from './styles';

type Props = Record<string, never>;

const Notification: FunctionComponent<Props> = ({}: Props) => {
    return (
        <Wrapper>
            <Sidebar />

            <MainContent>
                <Header />
                <NotificationWrapper>
                    <Title>Notifications</Title>
                    <Card>
                        <Container>
                            <NotificationList>
                                <NotificationItem />
                                <NotificationItem />
                                <NotificationItem />
                                <NotificationItem />
                                <NotificationItem />
                                <NotificationItem />
                                <NotificationItem />
                                <NotificationItem />
                                <NotificationItem />
                                <NotificationItem />
                                <NotificationItem />
                                <NotificationItem />
                                <NotificationItem />
                                <NotificationItem />
                                <NotificationItem />
                            </NotificationList>
                            <NotificationContent>
                                <PlaceholderContainer>
                                    <PlaceholderFox
                                        src={FoxImage}
                                        alt="Fox Mascot"
                                    />
                                    <PlaceholderText>
                                        Press to anybody to start chat with them
                                    </PlaceholderText>
                                </PlaceholderContainer>
                            </NotificationContent>
                        </Container>
                    </Card>
                </NotificationWrapper>
            </MainContent>
        </Wrapper>
    );
};

export default Notification;
