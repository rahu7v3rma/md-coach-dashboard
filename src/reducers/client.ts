import {
    createAsyncThunk,
    createEntityAdapter,
    createSlice
} from '@reduxjs/toolkit';
import moment from 'moment';
import { SortOrder } from 'react-data-table-component';
import { useSelector } from 'react-redux';

import * as api from '../services/api';
import { RootState } from '../store';
import { Colors } from '../utils/colors';
import { ClientProfiles } from 'src/types/client';
import { logBooks } from 'src/types/log';
import { Note, Notes } from 'src/types/note';
import { UserClient, UserGroup } from 'src/types/user';

const clientAdapter = createEntityAdapter();

interface ClientState {
    loading: boolean;
    clients: {
        list: Array<UserClient>;
        searchInput: string;
        hasPrevious: boolean;
        hasNext: boolean;
        totalCount: number;
        nextPageNumber: number;
        previousPageNumber: number;
        sortBy: number;
        order: SortOrder;
    };
    profiles: ClientProfiles;
    logBook: logBooks;
    notes: Notes;
    groups: {
        list: Array<UserGroup>;
        searchInput: string;
        hasPrevious: boolean;
        hasNext: boolean;
        totalCount: number;
        nextPageNumber: number;
        previousPageNumber: number;
        sortBy: number;
        order: SortOrder;
    };
    analytics: any;
}

const initialState = clientAdapter.getInitialState({
    loading: false,
    profiles: {},
    logBook: {},
    notes: {},
    clients: {
        list: [],
        searchInput: '',
        hasPrevious: false,
        hasNext: false,
        totalCount: 0,
        nextPageNumber: 1,
        previousPageNumber: 0,
        sortBy: 0,
        order: 'asc'
    },
    groups: {
        list: [],
        searchInput: '',
        hasPrevious: false,
        hasNext: false,
        totalCount: 4,
        nextPageNumber: 1,
        previousPageNumber: 0,
        sortBy: 0,
        order: 'asc'
    },
    analytics: []
} as ClientState);

export const getUserInfo = createAsyncThunk(
    'client/getUserInfo',
    async (id: string, { rejectWithValue }) => {
        try {
            const userInfo = await api.getUserInfo(id);
            return {
                ...userInfo,
                id
            };
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

export const getUserAnalytics = createAsyncThunk(
    'client/getUserAnalytics',
    async (
        { id, startDate, endDate }: Record<string, string | number>,
        { rejectWithValue }
    ) => {
        try {
            return await api.getUserAnalytics(
                id as string | number,
                startDate as string,
                endDate as string
            );
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

export const getClients = createAsyncThunk(
    'client/customers',
    async (
        { page, limit, search, sortBy, order }: Record<string, string | number>,
        { rejectWithValue }
    ) => {
        try {
            return await api.getClients(
                page as number,
                limit as number,
                search as string,
                sortBy as number,
                order as SortOrder
            );
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

export const getLogBook = createAsyncThunk(
    'user/getLogBook',
    async (
        options: {
            page: number;
            limit: number;
            user_id: string;
            category: string;
        },
        { rejectWithValue }
    ) => {
        try {
            return await api.getLogBook(
                options.page,
                options.limit,
                options.user_id,
                options.category
            );
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

export const onSearchChange = createAsyncThunk(
    'onSearchChange',
    async (search: string) => {
        return search;
    }
);

export const getNotes = createAsyncThunk(
    'client/getNotes',
    async (options: { page: number; limit: number; clientId: string }) => {
        return await api.getNotes(
            options.page,
            options.limit,
            options.clientId
        );
    }
);

export const addNote = createAsyncThunk(
    'client/addNote',
    async (options: {
        id: number;
        coach_id?: string;
        client_id: string;
        content: string;
        created_at: Date;
    }): Promise<Note> => {
        return await api.addNote(+options.client_id, options.content);
    }
);

export const editNote = createAsyncThunk(
    'client/editNote',
    async (options: {
        id: number;
        coach_id?: string;
        client_id: string;
        content: string;
        created_at: Date;
    }): Promise<Note> => {
        return await api.updateNote(
            options.id,
            +options.client_id,
            options.content
        );
    }
);

export const deleteNote = createAsyncThunk(
    'client/deleteNote',
    async (options: { id: number; clientId: string }) => {
        return await api.deleteNote(options.id);
    }
);

const clientSlice: any = createSlice({
    name: 'client',
    initialState,
    reducers: {
        sortOrder: (state, action) => {
            state.clients.sortBy = action.payload.sortBy;
            state.clients.order = action.payload.order;
        },
        sortGroupOrder: (state, action) => {
            state.groups.sortBy = action.payload.sortBy;
            state.groups.order = action.payload.order;
        }
    },
    extraReducers: (builder) => {
        builder.addCase(getLogBook.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(getLogBook.fulfilled, (state, action) => {
            state.loading = false;
            const userID = action.meta.arg.user_id;
            if (action.payload.has_previous) {
                const old_list = state.logBook[userID].list;
                const appended_list = [...old_list, ...action.payload.list];
                action.payload.list = appended_list;
                state.logBook[userID] = action.payload;
            } else {
                state.logBook[userID] = action.payload;
            }
        });
        builder.addCase(getLogBook.rejected, (state) => {
            state.loading = false;
        });
        builder.addCase(getUserInfo.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(getUserInfo.fulfilled, (state, action) => {
            state.loading = false;

            const profiles: ClientProfiles = { ...state.profiles };
            profiles[action.payload.id] = action.payload;

            state.profiles = profiles;
        });
        builder.addCase(getUserInfo.rejected, (state) => {
            state.loading = false;
        });
        builder.addCase(getClients.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(getClients.fulfilled, (state, action) => {
            state.loading = false;
            state.clients.hasPrevious = action.payload.has_previous;
            state.clients.hasNext = action.payload.has_next;
            state.clients.nextPageNumber = action.payload.next_page_number;
            state.clients.totalCount = action.payload.count;
            state.clients.list = action.payload.list;
            state.clients.sortBy = action.meta.arg.sortBy as number;
            state.clients.order = action.meta.arg.order as SortOrder;
        });
        builder.addCase(getClients.rejected, (state) => {
            state.loading = false;
        });
        builder.addCase(onSearchChange.fulfilled, (state, action) => {
            state.clients.searchInput = action.payload;
        });
        builder.addCase(getNotes.fulfilled, (state, action) => {
            const userID = action.meta.arg.clientId;
            if (action.payload.has_previous) {
                const old_list = state.notes[userID].list;
                const appended_list = [...old_list, ...action.payload.list];
                action.payload.list = appended_list;
                state.notes[userID] = action.payload;
            } else {
                state.notes[userID] = action.payload;
            }
        });
        builder.addCase(addNote.fulfilled, (state, action) => {
            state.notes[action.meta.arg.client_id].list.push(action.payload);
        });
        builder.addCase(editNote.fulfilled, (state, action) => {
            let previous_notes = {
                ...state.notes[action.meta.arg.client_id]
            };
            let edited_note = previous_notes.list.filter(
                (note) => note.id === action.meta.arg.id
            );
            if (edited_note.length) {
                edited_note[0].content = action.payload.content;
            }
            state.notes[action.meta.arg.client_id] = previous_notes;
        });
        builder.addCase(deleteNote.fulfilled, (state, action) => {
            let previous_notes = {
                ...state.notes[action.meta.arg.clientId]
            };
            previous_notes.list = previous_notes.list.filter(
                (note) => note.id !== action.meta.arg.id
            );
            state.notes[action.meta.arg.clientId] = previous_notes;
        });
        builder.addCase(getUserAnalytics.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(getUserAnalytics.rejected, (state) => {
            state.loading = false;
        });
        builder.addCase(getUserAnalytics.fulfilled, (state, action: any) => {
            const UserFastData: any =
                Object.keys(action.payload.UserFast).map((key: string) => ({
                    x: moment(key).format('DD/MM/YY'),
                    y:
                        action.payload.UserFast[key]?.length > 0
                            ? action.payload.UserFast[key][0]
                                  ?.duration_minutes ?? 0
                            : 0,
                    unit: 'min'
                })) ?? [];
            const UserGlucoseData: any =
                Object.keys(action.payload.UserGlucose).map((key: string) => ({
                    x: moment(key).format('DD/MM/YY'),
                    y:
                        action.payload.UserGlucose[key]?.length > 0
                            ? action.payload.UserGlucose[key][0]?.amount ?? 0
                            : 0,
                    unit:
                        action.payload.UserGlucose[key]?.length > 0
                            ? action.payload.UserGlucose[key][0]?.unit ?? ''
                            : ''
                })) ?? [];
            const UserWeightData: any =
                Object.keys(action.payload.UserWeight).map((key: string) => ({
                    x: moment(key).format('DD/MM/YY'),
                    y:
                        action.payload.UserWeight[key]?.length > 0
                            ? Math.round(
                                  action.payload.UserWeight[key][0]?.amount
                              ) ?? 0
                            : 0,
                    unit:
                        action.payload.UserWeight[key]?.length > 0
                            ? action.payload.UserWeight[key][0]?.unit ?? ''
                            : ''
                })) ?? [];
            const UserInsulinData: any =
                Object.keys(action.payload.UserInsulin).map((key: string) => ({
                    x: moment(key).format('DD/MM/YY'),
                    y:
                        action.payload.UserInsulin[key]?.length > 0
                            ? action.payload.UserInsulin[key][0]?.units ?? 0
                            : 0,
                    unit:
                        action.payload.UserInsulin[key]?.length > 0
                            ? action.payload.UserInsulin[key][0]
                                  ?.injection_type ?? ''
                            : ''
                })) ?? [];

            state.analytics = [
                {
                    title: 'Glucose',
                    color: Colors.extra.darkTerraCotta,
                    chart: {
                        type: 'line' as 'line',
                        data: UserGlucoseData
                    }
                },
                {
                    title: 'Weight (kg)',
                    color: Colors.extra.skyBlue2,
                    chart: {
                        type: 'line' as 'line',
                        data: UserWeightData
                    }
                },
                {
                    title: 'Fasting',
                    color: Colors.extra.lightPink,
                    chart: {
                        type: 'bar' as 'bar',
                        data: UserFastData
                    }
                },
                {
                    title: 'Insulin',
                    color: Colors.extra.lightGreen2,
                    chart: {
                        type: 'bar' as 'bar',
                        data: UserInsulinData
                    }
                }
            ];
        });
    }
});

export default clientSlice.reducer;

interface ClientSelectorsType {
    loading: boolean;
    logBook: any;
    notes: Notes;
    profiles: ClientProfiles;
    clients: Array<UserClient>;
    searchInput: string;
    hasPrevious: boolean;
    hasNext: boolean;
    totalCount: number;
    totalGroupCount: number;
    nextPageNumber: number;
    previousPageNumber: number;
    sortBy: number;
    order: string;
    groups: Array<UserGroup>;
    groupSortBy: number;
    groupOrder: string;
    analytics: any;
}

export const { sortOrder, sortGroupOrder } = clientSlice.actions;

export const ClientSelectors = (): ClientSelectorsType => {
    const loading = useSelector((state: RootState) => state.client.loading);
    const clients = useSelector(
        (state: RootState) => state.client.clients.list
    );
    const groups = useSelector((state: RootState) => state.client.groups.list);
    const hasPrevious = useSelector(
        (state: RootState) => state.client.clients.hasPrevious
    );
    const hasNext = useSelector(
        (state: RootState) => state.client.clients.hasNext
    );
    const totalGroupCount = useSelector(
        (state: RootState) => state.client.groups.totalCount
    );
    const totalCount = useSelector(
        (state: RootState) => state.client.clients.totalCount
    );
    const searchInput = useSelector(
        (state: RootState) => state.client.clients.searchInput
    );
    const nextPageNumber = useSelector(
        (state: RootState) => state.client.clients.nextPageNumber
    );
    const previousPageNumber = useSelector(
        (state: RootState) => state.client.clients.previousPageNumber
    );

    const sortBy = useSelector(
        (state: RootState) => state.client.clients.sortBy
    );

    const order = useSelector((state: RootState) => state.client.clients.order);

    const groupSortBy = useSelector(
        (state: RootState) => state.client.groups.sortBy
    );

    const groupOrder = useSelector(
        (state: RootState) => state.client.groups.order
    );

    const profiles = useSelector((state: RootState) => state.client.profiles);

    const logBook = useSelector((state: RootState) => state.client.logBook);

    const notes = useSelector((state: RootState) => state.client.notes);

    const analytics = useSelector((state: RootState) => state.client.analytics);

    return {
        loading,
        logBook,
        notes,
        profiles,
        clients,
        searchInput,
        hasPrevious,
        hasNext,
        totalCount,
        nextPageNumber,
        previousPageNumber,
        sortBy,
        order,
        groups,
        totalGroupCount,
        groupSortBy,
        groupOrder,
        analytics
    };
};
