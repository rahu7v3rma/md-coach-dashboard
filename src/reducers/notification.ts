import {
    createAsyncThunk,
    createEntityAdapter,
    createSlice
} from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import * as api from '../services/api';
import { RootState } from 'src/store';
import { NotificationListResponse } from 'src/types/notification';

interface NotificationState {
    loading: boolean;
    notifications: NotificationListResponse;
    unreadCount: number;
}

const userAdapter = createEntityAdapter();

const initialState = userAdapter.getInitialState({
    loading: false,
    notifications: {
        list: [],
        hasPrevious: false,
        hasNext: false,
        totalCount: 0,
        nextPageNumber: 1,
        previousPageNumber: 0
    },
    unreadCount: 0
} as NotificationState);

export const getNotification = createAsyncThunk(
    'user/notifications',
    async (
        options: {
            page: number;
            limit: number;
        },
        { rejectWithValue }
    ) => {
        try {
            return await api.getNotifction(options.page, options.limit);
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

export const readNotification = createAsyncThunk(
    'user/readNotifications',
    async (
        options: {
            id: number;
        },
        { rejectWithValue }
    ) => {
        try {
            return await api.readNotification(options.id);
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

const notificationSlice = createSlice({
    name: 'notification',
    initialState,
    reducers: {},
    extraReducers: (builder) => {
        builder.addCase(getNotification.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(getNotification.fulfilled, (state, action) => {
            state.loading = false;
            if (state.notifications.hasPrevious) {
                const oldData = state.notifications.list;
                action.payload.list = [...oldData, ...action.payload.list];
            }
            state.notifications.list = action.payload.list;
            state.unreadCount = action.payload.total_unread;
            state.loading = false;
        });
        builder.addCase(getNotification.rejected, (state) => {
            state.loading = false;
        });
        builder.addCase(readNotification.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(readNotification.fulfilled, (state, action) => {
            state.loading = false;
            state.notifications.list = state.notifications.list.map((item) => {
                if (item.id === action.meta.arg.id) {
                    item.read_flag = true;
                    state.unreadCount = state.unreadCount - 1;
                }
                return item;
            });
        });
        builder.addCase(readNotification.rejected, (state) => {
            state.loading = false;
        });
    }
});

export default notificationSlice.reducer;

interface NotificationSelectorsType {
    loading: boolean;
    notifications: NotificationListResponse;
    unreadCount: number;
}

export const NotificationSelectors = (): NotificationSelectorsType => {
    const loading = useSelector(
        (state: RootState) => state.notification.loading
    );

    const notifications = useSelector(
        (state: RootState) => state.notification.notifications
    );

    const unreadCount = useSelector(
        (state: RootState) => state.notification.unreadCount
    );
    return {
        loading,
        notifications,
        unreadCount
    };
};
