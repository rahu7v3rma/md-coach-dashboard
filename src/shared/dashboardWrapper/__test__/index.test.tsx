import React from 'react';
import { Provider } from 'react-redux';
import * as router from 'react-router';
import { BrowserRouter } from 'react-router-dom';
import { act, create } from 'react-test-renderer';

import store from '../../../store';
import * as notification from 'src/reducers/notification';
import DashboardLayout from 'src/shared/dashboardWrapper';
import Header from 'src/shared/header';

const navigate = jest.fn();
const mockGetNotification = jest.fn();
jest.mock('src/reducers/notification', () => ({
    NotificationSelectors: jest
        .fn()
        .mockImplementationOnce(() => ({
            loading: false,
            notifications: {
                hasNext: false
            },
            unreadCount: 0
        }))
        .mockImplementation(() => ({
            loading: true,
            notifications: {
                hasNext: true,
                nextPageNumber: 1
            },
            unreadCount: 0
        })),
    getNotification: jest.fn().mockImplementation(() => mockGetNotification())
}));

describe('DashboardLayout', () => {
    beforeEach(() => {
        jest.spyOn(router, 'useNavigate').mockImplementation(() => navigate);
        jest.clearAllMocks();
    });

    // notification selector mock once with hasNext false so no more notifications are there
    it('handleScroll does not dispatch notification retrieval action when there are no more notifications', async () => {
        const tree = create(
            <Provider store={store}>
                <BrowserRouter>
                    <DashboardLayout>
                        <div id="child" />
                    </DashboardLayout>
                </BrowserRouter>
            </Provider>
        );
        let header = tree.root.findByType(Header).props;
        const getNotification = jest.spyOn(notification, 'getNotification');
        await act(() => header.handleScroll());
        expect(getNotification).not.toHaveBeenCalled();
    });

    // notification selector mock with hasNext true so more notifications are there
    it('handleScroll dispatches notification retrieval action when there are more notifications', async () => {
        const tree = create(
            <Provider store={store}>
                <BrowserRouter>
                    <DashboardLayout>
                        <div id="child" />
                    </DashboardLayout>
                </BrowserRouter>
            </Provider>
        );
        let header = tree.root.findByType(Header).props;
        await act(() => header.handleScroll());
        expect(mockGetNotification).toHaveBeenCalled();
    });

    it('DashboardLayout snapshot', async () => {
        const tree = create(
            <Provider store={store}>
                <BrowserRouter>
                    <DashboardLayout />
                </BrowserRouter>
            </Provider>
        );
        expect(tree).toMatchSnapshot();
    });

    it('DashboardLayout renders without errors', async () => {
        const tree = create(
            <Provider store={store}>
                <BrowserRouter>
                    <DashboardLayout />
                </BrowserRouter>
            </Provider>
        );
        expect(tree.root).toBeTruthy();
    });

    it('DashboardLayout renders its children', async () => {
        const tree = create(
            <Provider store={store}>
                <BrowserRouter>
                    <DashboardLayout>
                        <div id="child" />
                    </DashboardLayout>
                </BrowserRouter>
            </Provider>
        );
        let child = tree.root.findByProps({
            id: 'child'
        }).props;
        expect(child).toBeTruthy();
    });

    it('handleToggle toggles the "hidden" state', async () => {
        const tree = create(
            <Provider store={store}>
                <BrowserRouter>
                    <DashboardLayout>
                        <div id="child" />
                    </DashboardLayout>
                </BrowserRouter>
            </Provider>
        );
        let header = tree.root.findByType(Header).props;
        expect(header.hidden).toBe(true);
        await act(() => header.handleToggle());
        let headerHide = tree.root.findByType(Header).props;
        expect(headerHide.hidden).toBe(false);
    });
});
