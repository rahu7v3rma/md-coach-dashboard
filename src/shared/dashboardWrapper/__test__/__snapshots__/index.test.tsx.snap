// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardLayout DashboardLayout snapshot 1`] = `
<div
  className="sc-hfKJst kRMkjY"
>
  <div
    className="sc-dyTUbJ fzBoHW"
  >
    <div
      className="sc-kJpdyg gaapxs"
    >
      <div
        className="sc-kJwmPe jHHKdK"
      >
        <img
          alt="MD Icon"
          className="sc-emqRaN jhIXvV"
          src={Object {}}
        />
        <p
          className="sc-dzbdsH gMHdzK"
        >
          Mastering 
          <br />
          Programs
        </p>
      </div>
      <div
        className="sc-dmovaM eTRJKf"
      >
        <div
          className="sc-eWhHU hgAbBM"
          onClick={[Function]}
          selected={false}
        >
          <div
            className="sc-brSOsn dSyhQB"
          >
            <img
              alt="My Clients"
              src={Object {}}
            />
          </div>
          <p
            className="sc-iCnXOK ewXqqL"
            selected={false}
          >
            My Clients
          </p>
        </div>
        <div
          className="sc-eWhHU hgAbBM"
          onClick={[Function]}
          selected={false}
        >
          <div
            className="sc-brSOsn dSyhQB"
          >
            <img
              alt="Chat"
              src={Object {}}
            />
          </div>
          <p
            className="sc-iCnXOK ewXqqL"
            selected={false}
          >
            Chat
          </p>
        </div>
      </div>
      <p
        className="sc-ePzlA-D ifuDMT sc-dSbvwS jLFnCT"
        color="#B16663"
        fontSize={12}
        fontWeight="500"
        onClick={[Function]}
      >
        Notifications are disabled. Click here to grant required permissions
      </p>
    </div>
    <div
      className="sc-ccVCaX fJHeJO"
    >
      <div
        className="sc-jnqLxu hHovem"
      >
        <div
          className="sc-ejWHKt fmxHYX"
        >
          <p
            className="sc-ePzlA-D lbcyHB"
            fontWeight="600"
          >
            Logout
          </p>
          <p
            className="sc-ePzlA-D jSYaFQ"
            color="#A4AAAF"
            fontSize={11}
            fontWeight="500"
          >
            Click below to logout the dashboard
          </p>
        </div>
        <button
          className="sc-dcrmVg kotmLE"
          height="40px"
          id="logoutBtn"
          onClick={[Function]}
        >
          <p
            className="sc-ePzlA-D fLRZEA"
            fontSize={12}
            fontWeight="600"
          >
            Logout
          </p>
        </button>
      </div>
    </div>
  </div>
  <div
    className="sc-jjRTzg gslEFu"
  >
    <div
      className="sc-iRQbrd eeIqON"
    >
      <div
        className="sc-XMxGW hEIGnj"
      >
        <div
          className="sc-kPTPQs jiieSE"
        >
          <img
            className="sc-ispOId eHznbG"
            src={Object {}}
          />
          <input
            className="sc-kcuKUB LwXuE"
            onChange={[Function]}
            placeholder="Search"
          />
        </div>
      </div>
      <div
        className="sc-TJdGy cjOwfa"
      >
        <div
          className="sc-eojhpR dAMxmL"
          onClick={[Function]}
        >
          <img
            className="sc-jlRNHb dKYKeX"
            src={Object {}}
          />
        </div>
        <div
          className="sc-fFGjHI fiLKAe"
        >
          <label>
            <input
              accept="image/png, image/jpeg"
              aria-label="File input"
              className="sc-gxYJeL fcIGhX"
              id="handleSelectedFileChanged"
              onChange={[Function]}
              type="file"
            />
            <div
              className="sc-dicizt hLxdUU sc-DooS jRKZHN"
              id="editOverlay"
            >
              <img
                alt="avatar"
                className="sc-ewnqHT sc-DooS jRKZHN"
                id="defaultAvatar"
                src={Object {}}
              />
              <img
                alt="edit"
                className="sc-iVCKna dSjaUI"
                src={Object {}}
              />
            </div>
          </label>
        </div>
      </div>
      <div
        className="sc-ffyeCz hIHYdN"
        hidden={true}
        onScroll={[Function]}
      >
        <div
          className="sc-flsqHV hKdGbz"
        >
          <span
            className="sc-hpzAOI TMXRk"
          >
            Notifications
          </span>
          <img
            alt="Bell Icon"
            className="sc-kjaWZf bTwOLq"
            onClick={[Function]}
            role="button"
            src={Object {}}
          />
        </div>
      </div>
    </div>
    <div
      className="sc-bgrQFw gKErPa"
    />
  </div>
</div>
`;
